Flex Attention HOP: Add support for flex decoding

joydddd opened the pull request:
# Flex Decoding
tl;dr This PR adds `flex_decoding` kernel to higher-order-op: `flex_attention` as the backend for multi-head attention decoding. 

Higher-order-op `flex_attention` was introduced in (https://github.com/pytorch/pytorch/pull/121845) to accept a user defined score modification callable (`score_mod`) and through `torch.compile`to create an efficient fused flash attention kernel instatiation. The `flex_attention` kernel is efficient for long queries (>512 tokens) attention. This PR introduces `flex_decoding` kernel as an alternative backend for `flex_attention` HOP to handle LLM inference where short queries (<32 tokens) attends to long key/value sequences.

### Details

LLM decoding iteratively attends each newly generated token ( query length = 1 ) to a long key/value context (up to 132k). `flex_attention` kernel only parallelizes attention along query length (M), batch size (B) and number of heads (H) dimension. LLM decoding lacks enough parallelism in the M dimension to fill up all SMs on the modern GPUs. 

`flex_decoding` adds parallelization along key/value sequence length (N). The key/value cache of a single head are split into multiple blocks and the query tokens attends to them in parallel. The results for the same head are then reduced across KV blocks to generate a global output. 


## Examples

Consider a Group Query Attention (GQA) decoding case, where a query token of 16 query heads (Hq) attends to 2 kv head (Hkv). Assume a batch size of 2 (B=2) and kv cache length of 4096 (N=4096). The attention kernel iteratively attends to newly generated query token (Mq = 1). 

We transform this problem into a Multiheaded Attention (MHA) problem by assuming a query length equal to number of query heads per kv heads, i.e. M=Hq//Hkv. 
The inputs to `flex_attention` HOP is thus a query of shape (B=2, H=Hkv=2, M=Hq//Hkv=8, D=64), key,value of shape (B=2, H=Hkv=2, N=4096, D=64, which lead to an intermediate attention score matrix of shape (2, 2, 8, 4096) and an output of shape (2, 2, 8, 64). 

```Python
import torch
from torch.nn.attention._flex_attention import _flex_attention as flex_attention

torch.manual_seed(0)

# Lets create some input tensors
# query of shape (B, Hkv, Hq//Hkv, D) 
# key/value of shape (B, Hkv, N, D) 
query = torch.randn(2, 2, 8, 64, device="cuda", dtype=torch.float32)
key = torch.randn(2, 2, 4096, 64, device="cuda", dtype=torch.float32)
value = torch.randn(2, 2, 4096, 64, device="cuda", dtype=torch.float32)

# Lets create a new score_modification checkerboard. 
def checkerboard(score, batch, head, token_q, token_kv):
    score = torch.where(torch.abs(token_kv - token_q) == 1, score * 0.5, score)
    score = torch.where(torch.abs(token_kv - token_q) == 2, score * 2.0, score)
    return score

# Lets call flex_attention with this new score modification for decoding. 
# The flex_attention HOP will chose flex_decoding as its backend since our query length (M) is only 8. 
output = flex_attention(query, key, value, score_mod=checkerboard)

compiled_flex_attention = torch.compile(flex_attention)
out_compiled = compiled_flex_attention (query, key, value, score_mod=checkerboard)

torch.testing.assert_close(output, out_compiled, atol=2e-2, rtol=2e-2)
```

## Future Plans 
- This PR does not implement load mask for score_mod function. This means if the score_mod functions takes a captured buffer along the M dimension , it must be padded to q length of 16, or next 2^n of query length if q_len > 16. 
i.e.
```python
q_scale = torch.randn(Hq//Hkv, device="cuda")
q_scale = torch.nn.functional.pad(q_scale, (0, 16-Hq//Hkv)) # Pad captured buffer
def bias_mod(score, batch, head, q, kv):
    score = score + q_scale[token_q]
    return score
```
- Backward path for short queries (<128 token) currently does not work because the `flex_attention_backward` kernel is lacking mask support and only takes query length of a multiple of 128. 
- Dynamic shape and max_autotuning is currently not working
- Add block sparse mask support (#129216 is a draft for flex_attention kernel) 
- Add explicit GQA support. (#130076 is a draft for GQA support on flex_attention kernel) 




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

joydddd commented:
Needs Triton Fix https://github.com/triton-lang/triton/pull/4061

joydddd commented:
Use inductor generated fused reduction kernel. Performance is similar to hand written templated kernel.

joydddd commented:
Split_kv tuning: one block per SM. 


### Speedups w.r.t Xformers SplitK Kernel
~1 -> avg. 1.22x max 1.8x

| Type    |   Speedup | score_mod   | dtype         | shape(B,Hkv,Hq//Hkv,M,D,N)   |   BandWidth (TB/s) |   GFlops/s |
|---------|-----------|-------------|---------------|------------------------------|--------------------|------------|
| Average |     1.224 |             |               |                              |                    |            |
| Max     |     1.846 | noop        | torch.float16 | (2, 1, 16, 64, 131072)       |              0.983 |  17425.572 |
| Min     |     0.998 | noop        | torch.float16 | (16, 2, 8, 64, 8192)         |              0.953 |   8438.743 |

drisspg commented on test/inductor/test_flex_attention.py:
```diff
@@ -593,19 +633,63 @@ def func(q, k, v, k2, v2):
+        k1 = torch.randn((B * Hkv * S * D * 4), dtype=dtype, device="cuda")
+        v1 = torch.randn((B * Hkv * S * D * 4), dtype=dtype, device="cuda")
+
+        q_shape = (B, H, Hq // Hkv, D)
```
It might be more clear if (B, Hkv, Hq // hkv, D)

drisspg replied:
or maybe name some things, group_size

also since users would be expected to do this GQA like grouping, we might want to add a utility for doing this.

joydddd replied:
sorry, my bad. This q_shape isn't used in the code below.(B, Hkv, Hq//Hkv, D) is the shape actually used.

drisspg commented on test/inductor/test_flex_attention.py:
```diff
@@ -614,32 +698,37 @@ def score_mod_2(score, b, h, m, n):
+        kv_scale = torch.randn(S, device="cuda")
+        q_scale = torch.randn(Hq // Hkv if decoding else S, device="cuda")
+        if q_scale.shape[-1] < 16:
+            q_scale = torch.nn.functional.pad(q_scale, (0, 16 - Hq // Hkv))
```
If users would need to pad the buffers they are going to capture, maybe we should handle this during the lowering? this feels hard to to use

joydddd replied:
The ultimate solution is adding load mask to modification{{}} generated code. Like this: 
https://github.com/joydddd/pytorch/pull/2/files# But we have some performance problems.  

Just a temp work around before we figure that out.

Chillee commented on torch/_inductor/kernel/flex_decoding.py:
```diff
@@ -0,0 +1,363 @@
+    if SAFE_M_BOUNDARY:
+        q = tl.load(Q_block_ptr)
+    else:
+        q = tl.load(Q_block_ptr, boundary_check=(0, ))
```
Does this impact perf?

Chillee commented on torch/_inductor/kernel/flex_decoding.py:
```diff
@@ -0,0 +1,363 @@
+    idx_m = offs_m[:, None]
+    idx_d = offs_d[None, :]
+    # TODO generalize and add proper mask support
+    mask = (idx_m < Q_CTX) & (idx_d != -1)
```
I think you only need `mask = (idx_m < Q_CTX)`.

Chillee commented on torch/_inductor/kernel/flex_decoding.py:
```diff
@@ -0,0 +1,363 @@
+    Mk = key.get_size()[-2]
+    head_dim = key.get_size()[-1]
+    dtype = key.get_dtype()
+    default_config = None
```
This function seems to havea  lot of dead code?

Chillee commented on torch/_inductor/kernel/flex_decoding.py:
```diff
@@ -0,0 +1,363 @@
+
+    BLOCK_M = (
+        query.get_size()[-2]
+        if query.get_size()[-2] <= 0
```
add a note that this is intentional.

joydddd replied:
This is a works around before Triton fix https://github.com/triton-lang/triton/pull/4061 is in pin.

Chillee reviewed (changes requested):
I would prefer to have a separate `test_flex_decoding.py` test file for now, i think. There's too many changes and various workarounds for tests here.

In general, I think that I generally lean towards making tests easy to understand what they're testing (and be fine with duplication).

joydddd commented on test/inductor/test_flex_attention.py:
```diff
@@ -547,7 +587,7 @@ def func(q, k, v):
         # Ensure _create_block_mask_from_mask is compiled and generates 3 kernels,
         # flex_attention generates 1 kernel.
-        FileCheck().check_count(".run(", 4, True).run(code[0])
+        FileCheck().check_count(".run(", 4, False).run(code[0])
```
To allow lowering of mask generation generate different number of kernels. It was failing in trunk. We might need to a way to better test this.

joydddd commented on test/inductor/test_flex_attention.py:
```diff
@@ -217,7 +227,15 @@ def run_test(
         KV_H: int = H,
         KV_S: int = S,
         KV_D: int = D,
+        Hq: int = Hq,
```
Flex_decoding in a new file now.

joydddd committed:
Rebase flex_dec to main

fix lint

Add review suggestions

Add scale factor to flex_dec

Rename flex dec kernel var to follow flex att

joydddd commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -370,6 +375,51 @@ def build_subgraph_buffer(
+    return V.graph.sizevars.evaluate_expr(sympy.Lt(query.get_size()[-2], 32 + 1))
+
+
+def filtered_configs(
```
I would like to handle decoding cases where q seq len = 64 by flex_attention. After some performance benchmarking, I think it makes sense to direct them to decoding as well. This filter is now removed. 

Eventually we probably would need a better way to decide between kernels?

joydddd committed:
Add test for backward flex_decoding fails

joydddd commented:
@pytorch-bot merge

joydddd commented:
@pytorchbot merge

joydddd commented:
@pytorchbot merge

joydddd commented:
@pytorchbot label "topic: not user facing"

joydddd commented:
@pytorchbot merge
